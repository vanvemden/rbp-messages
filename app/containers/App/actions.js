/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  GET_MESSAGES,
  GET_MESSAGES_SUCCESS,
  GET_MESSAGES_ERROR,
  POST_MESSAGE,
  POST_MESSAGE_SUCCESS,
  POST_MESSAGE_ERROR,
  PATCH_MESSAGE,
  PATCH_MESSAGE_SUCCESS,
  PATCH_MESSAGE_ERROR,
  DELETE_MESSAGE,
  DELETE_MESSAGE_SUCCESS,
  DELETE_MESSAGE_ERROR,
} from './constants';

/**
 * Get the messages, this action starts the request saga. Saga will dispatch:
 * - on success, getMessagesSuccess(messages)
 * - on error, getMessagesError(error)
 */
export function getMessages() {
  return {
    type: GET_MESSAGES,
  };
}

export function getMessagesSuccess(messages) {
  return {
    type: GET_MESSAGES_SUCCESS,
    messages,
  };
}

export function getMessagesError(error) {
  return {
    type: GET_MESSAGES_ERROR,
    error,
  };
}

/**
 * Post a message, this action starts the request saga. Saga will dispatch:
 * - on success, postMessageSuccess(message)
 * - on error, postMessageError(error)
 */
export function postMessage() {
  return {
    type: POST_MESSAGE,
  };
}

export function postMessageSuccess(message) {
  return {
    type: POST_MESSAGE_SUCCESS,
    message,
  };
}

export function postMessageError(error) {
  return {
    type: POST_MESSAGE_ERROR,
    error,
  };
}

/**
 * Patch a message, this action starts the request saga. Saga will dispatch:
 * - on success, patchMessageSuccess(message)
 * - on error, patchMessageError(error)
 */
export function patchMessage() {
  return {
    type: PATCH_MESSAGE,
  };
}

export function patchMessageSuccess(message) {
  return {
    type: PATCH_MESSAGE_SUCCESS,
    message,
  };
}

export function patchMessageError(error) {
  return {
    type: PATCH_MESSAGE_ERROR,
    error,
  };
}

/**
 * Delete a message, this action starts the request saga. Saga will dispatch:
 * - on success, deleteMessageSuccess(id)
 * - on error, deleteMessageError(error)
 */
export function deleteMessage() {
  return {
    type: DELETE_MESSAGE,
  };
}

export function deleteMessageSuccess(id) {
  return {
    type: DELETE_MESSAGE_SUCCESS,
    id,
  };
}

export function deleteMessageError(error) {
  return {
    type: DELETE_MESSAGE_ERROR,
    error,
  };
}
