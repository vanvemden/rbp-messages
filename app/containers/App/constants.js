/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const GET_POSTS = 'rbp-messages/App/GET_POSTS';
export const GET_POSTS_SUCCESS = 'rbp-messages/App/GET_POSTS_SUCCESS';
export const GET_POSTS_ERROR = 'rbp-messages/App/GET_POSTS_ERROR';

export const PATCH_POST = 'rbp-messages/App/PATCH_POST';
export const PATCH_POST_SUCCESS = 'rbp-messages/App/PATCH_POST_SUCCESS';
export const PATCH_POST_ERROR = 'rbp-messages/App/PATCH_POST_ERROR';

export const DELETE_POST = 'rbp-messages/App/DELETE_POST';
export const DELETE_POST_SUCCESS = 'rbp-messages/App/DELETE_POST_SUCCESS';
export const DELETE_POST_ERROR = 'rbp-messages/App/DELETE_POST_ERROR';
