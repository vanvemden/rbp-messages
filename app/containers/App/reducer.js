/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

import produce from 'immer';
import {
  GET_POSTS,
  GET_POSTS_SUCCESS,
  GET_POSTS_ERROR,
  POST_POST,
  POST_POST_SUCCESS,
  POST_POST_ERROR,
  PATCH_POST,
  PATCH_POST_SUCCESS,
  PATCH_POST_ERROR,
  DELETE_POST,
  DELETE_POST_SUCCESS,
  DELETE_POST_ERROR,
} from './constants';

// The initial state of the App
export const initialState = {
  loading: false,
  storing: false,
  error: false,
  posts: [],
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    let index;
    switch (action.type) {
      case GET_POSTS:
        draft.loading = true;
        draft.error = false;
        draft.posts = [];
        break;

      case GET_POSTS_SUCCESS:
        draft.posts = action.posts;
        draft.loading = false;
        break;

      case GET_POSTS_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;

      case POST_POST:
        draft.storing = true;
        draft.error = false;
        break;

      case POST_POST_SUCCESS:
        draft.posts.push(action.post);
        draft.storing = false;
        break;

      case POST_POST_ERROR:
        draft.error = action.error;
        draft.storing = false;
        break;

      case PATCH_POST:
        draft.storing = true;
        draft.error = false;
        break;

      case PATCH_POST_SUCCESS:
        index = draft.posts.findIndex(post => post.id === action.post.id);
        if (index !== -1) draft.posts[index] = action.post;
        draft.storing = false;
        break;

      case PATCH_POST_ERROR:
        draft.error = action.error;
        draft.storing = false;
        break;

      case DELETE_POST:
        draft.storing = true;
        draft.error = false;
        break;

      case DELETE_POST_SUCCESS:
        draft.posts.filter(post => post.id !== action.id);
        draft.storing = false;
        break;

      case DELETE_POST_ERROR:
        draft.error = action.error;
        draft.storing = false;
        break;
    }
  });

export default appReducer;
