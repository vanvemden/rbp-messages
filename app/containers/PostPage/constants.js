export const POST_POST = 'rbp-messages/PostPage/POST_POST';
export const POST_POST_SUCCESS = 'rbp-messages/PostPage/POST_POST_SUCCESS';
export const POST_POST_ERROR = 'rbp-messages/PostPage/POST_POST_ERROR';
