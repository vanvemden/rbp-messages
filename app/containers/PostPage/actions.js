import { POST_POST, POST_POST_SUCCESS, POST_POST_ERROR } from './constants';

/**
 * Post a post, this action starts the request saga. Saga will dispatch:
 * - on success, postPostSuccess(post)
 * - on error, postPostError(error)
 */
export function postPost(post) {
  return {
    type: POST_POST,
    post,
  };
}

export function postPostSuccess(post) {
  return {
    type: POST_POST_SUCCESS,
    post,
  };
}

export function postPostError(error) {
  return {
    type: POST_POST_ERROR,
    error,
  };
}
