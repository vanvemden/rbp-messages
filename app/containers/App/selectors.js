/**
 * The global state selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const selectRouter = state => state.router;

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectStoring = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.storing,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectPosts = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.posts,
  );

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

export {
  selectGlobal,
  makeSelectLoading,
  makeSelectStoring,
  makeSelectError,
  makeSelectPosts,
  makeSelectLocation,
};
