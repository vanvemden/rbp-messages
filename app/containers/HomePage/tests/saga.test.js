/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { GET_POSTS } from 'containers/App/constants';
import { getPostsSuccess, getPostsError } from 'containers/App/actions';

import apiGetPosts, { requestPosts } from '../saga';

const username = 'mxstbr';

/* eslint-disable redux-saga/yield-effects */
describe('requestPosts Saga', () => {
  let requestPostsGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    requestPostsGenerator = requestPosts();

    const selectDescriptor = requestPostsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = requestPostsGenerator.next(username).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  // TODO
  xit('should dispatch the getPostsSuccess action if it requests the data successfully', () => {
    const response = [
      {
        name: 'First repo',
      },
      {
        name: 'Second repo',
      },
    ];
    const putDescriptor = requestPostsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(getPostsSuccess(response, username)));
  });

  // TODO
  xit('should call the getPostsError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = requestPostsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(getPostsError(response)));
  });
});

describe('apiGetPostsSaga Saga', () => {
  const apiGetPostsSaga = apiGetPosts();

  it('should start task to watch for GET_POSTS action', () => {
    const takeLatestDescriptor = apiGetPostsSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(GET_POSTS, requestPosts));
  });
});
