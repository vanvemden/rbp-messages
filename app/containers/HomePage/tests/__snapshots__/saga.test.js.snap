// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestPosts Saga should call the getPostsError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "posts": Array [
        Object {
          "id": 1,
          "text": "Hello World!",
        },
        Object {
          "id": 2,
          "text": "How are you?",
        },
      ],
      "type": "rbp-messages/App/GET_POSTS_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`requestPosts Saga should call the getPostsError action if the response errors 2`] = `undefined`;

exports[`requestPosts Saga should dispatch the getPostsSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "posts": Array [
        Object {
          "id": 1,
          "text": "Hello World!",
        },
        Object {
          "id": 2,
          "text": "How are you?",
        },
      ],
      "type": "rbp-messages/App/GET_POSTS_SUCCESS",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`requestPosts Saga should dispatch the getPostsSuccess action if it requests the data successfully 2`] = `undefined`;
